internal/models/linking.go

```diff
@@ -59,58 +72,61 @@ func DetermineAccountLinking(tx *storage.Connection, provider, sub string, email
 		}
 
 		return AccountLinkingResult{
-			Decision:      AccountExists,
-			User:          user,
-			Identities:    []*Identity{identity},
-			LinkingDomain: GetAccountLinkingDomain(provider),
+			Decision:       AccountExists,
+			User:           user,
+			Identities:     []*Identity{identity},
+			LinkingDomain:  GetAccountLinkingDomain(provider),
+			CandidateEmail: candidateEmail,
 		}, nil
 	} else if !IsNotFoundError(terr) {
 		return AccountLinkingResult{}, terr
 	}
 
-	// account does not exist, identity and user not immediately
-	// identifiable, look for similar identities based on email
-	var similarIdentities []*Identity
-	var similarUsers []*User
+	// the identity does not exist, so we need to check if we should create a new account
+	// or link to an existing one
 
-	if len(emails) > 0 {
-		if terr := tx.Q().Eager().Where("email ilike any (?)", emails).All(&similarIdentities); terr != nil {
+	// this is the linking domain for the new identity
+	candidateLinkingDomain := GetAccountLinkingDomain(provider)
+	if len(verifiedEmails) == 0 {
+		// if there are no verified emails, we always decide to create a new account
+		user, terr := IsDuplicatedEmail(tx, candidateEmail.Email, aud, nil)
+		if terr != nil {
 			return AccountLinkingResult{}, terr
 		}
-
-		if !strings.HasPrefix(provider, "sso:") {
-			// there can be multiple user accounts with the same email when is_sso_user is true
-			// so we just do not consider those similar user accounts
-			if terr := tx.Q().Eager().Where("email ilike any (?) and is_sso_user is false", emails).All(&similarUsers); terr != nil {
-				return AccountLinkingResult{}, terr
-			}
+		if user != nil {
+			candidateEmail.Email = ""
 		}
-	}
-
-	// TODO: determine linking behavior over phone too
-	if len(similarIdentities) == 0 && len(similarUsers) == 0 {
-		// there are no similar identities, clearly we have to create a new account
-
 		return AccountLinkingResult{
-			Decision:      CreateAccount,
-			LinkingDomain: GetAccountLinkingDomain(provider),
+			Decision:       CreateAccount,
+			LinkingDomain:  candidateLinkingDomain,
+			CandidateEmail: candidateEmail,
 		}, nil
 	}
 
-	// there are some similar identities, we now need to proceed in
-	// identifying whether this supposed new identity should be assigned to
-	// an existing user or to create a new user, according to the automatic
-	// linking rules
+	var similarIdentities []*Identity
+	var similarUsers []*User
+	// look for similar identities and users based on email
+	if terr := tx.Q().Eager().Where("email ilike any (?)", verifiedEmails).All(&similarIdentities); terr != nil {
```

Comment:
Please don't use `ilke` or `like`! It's a performance nightmare, indexes are not used. If you must, please re-create the index with `text_pattern_ops` so lookups are closer to O(logn) rather than O(n).